File: tests/metrics/annualized_volatility_test.py
from __future__ import annotations
import empyrical
import numpy as np
import pytest
from taflow.metrics.annualized_volatility import AnnualizedVolatility

@pytest.mark.parametrize(('returns', 'periods_per_year'), [(np.array([0.01, -0.01]), 252.0), (np.array([0.1, -0.2, 0.05]), 12.0), (np.zeros(32), 365.0), (np.linspace(-0.04, 0.05, 101), 52.0), (np.array([np.nan, 0.02, -0.03, np.nan, 0.01]), 252.0), (np.random.default_rng(20260811).normal(0.0004, 0.012, 513), 252.0)])
def test_annualized_volatility_matches_empyrical(returns: np.ndarray, periods_per_year: float) -> None:
    actual = AnnualizedVolatility(periods_per_year=periods_per_year).from_returns(returns).compute()
    expected = float(empyrical.annual_volatility(returns, annualization=periods_per_year))
    assert actual == pytest.approx(expected, rel=1e-12, abs=1e-14)

def test_annualized_volatility_input_methods_and_lifecycle_are_invariant() -> None:
    returns = np.array([0.1, -0.2, 0.05])
    expected = AnnualizedVolatility(periods_per_year=12.0).from_returns(returns).compute()
    equity = np.array([100.0, 110.0, 88.0, 92.4])
    pnl = np.array([10.0, -22.0, 4.4])
    assert AnnualizedVolatility(periods_per_year=12.0).from_equity(equity).compute() == pytest.approx(expected)
    assert AnnualizedVolatility(periods_per_year=12.0).from_pnl(pnl, initial_capital=100.0).compute() == pytest.approx(expected)
    assert AnnualizedVolatility(periods_per_year=12.0).from_log_returns(np.log1p(returns)).compute() == pytest.approx(expected)
    state = AnnualizedVolatility(periods_per_year=12.0).from_returns([])
    assert state.value is None
    assert state.append(returns[0]) is state
    assert state.value is None
    assert state.extend(returns[1:]) is state
    assert state.compute() == pytest.approx(expected)
    assert len(state) == len(returns)
    assert state.reset() is state
    assert len(state) == 0
    assert state.extend(returns).compute() == pytest.approx(expected)

def test_annualized_volatility_missing_constant_and_invalid_values() -> None:
    assert AnnualizedVolatility().from_returns([0.25]).compute() is None
    assert AnnualizedVolatility().from_returns([0.25, 0.25]).compute() == 0.0
    assert len(AnnualizedVolatility().from_returns([np.nan, 0.01, 0.02])) == 2
    with pytest.raises(ValueError):
        AnnualizedVolatility(nan_policy='raise').from_returns([np.nan])
    with pytest.raises(ValueError):
        AnnualizedVolatility().from_returns([np.inf])
    with pytest.raises(ValueError):
        AnnualizedVolatility(periods_per_year=0.0).from_returns([0.01, 0.02])

def test_annualized_volatility_requires_semantic_ingestion_before_append() -> None:
    with pytest.raises(ValueError):
        AnnualizedVolatility().append(0.01)

File: crates/taflow-metrics/src/metrics/annualized_volatility_test.rs
use super::annualized_volatility::AnnualizedVolatility;
use crate::{MetricInputKind, NanPolicy};

fn assert_close(actual: f64, expected: f64) {
    assert!((actual - expected).abs() <= 1e-14, "{actual} != {expected}");
}

#[test]
fn computes_sample_volatility_and_preserves_lifecycle() {
    let returns = [0.01, -0.02, 0.03, 0.00];
    let expected = 0.02081665999466133 * 252.0_f64.sqrt();
    let mut state = AnnualizedVolatility::new(252.0, NanPolicy::Omit).unwrap();
    state.from_returns(&[]).unwrap();

    assert_eq!(state.value(), None);
    assert_eq!(state.append(returns[0]).unwrap(), None);
    state.extend(&returns[1..3]).unwrap();
    assert_close(state.append(returns[3]).unwrap().unwrap(), expected);
    assert_close(state.compute().unwrap(), expected);
    assert_eq!(state.len(), returns.len());

    state.reset();
    assert!(state.is_empty());
    assert_eq!(state.value(), None);
    assert_close(state.extend(&returns).unwrap().unwrap(), expected);
}

#[test]
fn input_modes_produce_equivalent_volatility() {
    let returns = [0.10, -0.20, 0.05];
    let expected = {
        let mut state = AnnualizedVolatility::new(12.0, NanPolicy::Omit).unwrap();
        state.from_returns(&[]).unwrap();
        state.extend(&returns).unwrap().unwrap()
    };

    let mut equity = AnnualizedVolatility::new(12.0, NanPolicy::Omit).unwrap();
    equity.from_equity(&[]).unwrap();
    assert_close(
        equity.extend(&[100.0, 110.0, 88.0, 92.4]).unwrap().unwrap(),
        expected,
    );
    assert_eq!(equity.len(), 3);

    let mut pnl = AnnualizedVolatility::new(12.0, NanPolicy::Omit).unwrap();
    pnl.from_pnl(&[], 100.0).unwrap();
    assert_close(pnl.extend(&[10.0, -22.0, 4.4]).unwrap().unwrap(), expected);

    let log_returns = returns.map(f64::ln_1p);
    let mut logarithmic = AnnualizedVolatility::new(12.0, NanPolicy::Omit).unwrap();
    logarithmic.from_log_returns(&[]).unwrap();
    assert_close(logarithmic.extend(&log_returns).unwrap().unwrap(), expected);
}

#[test]
fn handles_missing_constant_and_minimum_samples() {
    let mut state = AnnualizedVolatility::new(252.0, NanPolicy::Omit).unwrap();
    state.from_returns(&[]).unwrap();
    state.extend(&[f64::NAN, 0.25]).unwrap();
    assert_eq!(state.len(), 1);
    assert_eq!(state.value(), None);
    state.append(0.25).unwrap();
    assert_eq!(state.value(), Some(0.0));
}

#[test]
fn rejects_invalid_periods_per_year() {
    for value in [0.0, -1.0, f64::NAN, f64::INFINITY] {
        assert!(AnnualizedVolatility::new(value, NanPolicy::Omit).is_err());
    }
}

#[test]
fn rejects_non_return_semantic_domains() {
    let mut unbound = AnnualizedVolatility::new(252.0, NanPolicy::Omit).unwrap();
    assert!(unbound.append(0.01).is_err());
}
